File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findHeight(TreeNode* root, bool &ans){
        if(root==NULL) return 0;

        int lh=findHeight(root->left, ans);
        int rh=findHeight(root->right, ans);
        if(abs(lh-rh)>1) ans=false;

        return 1+max(lh,rh);
    }

    bool isBalanced(TreeNode* root) {
        if(root==nullptr) return -1;
        bool ans=true;
        int lh=0, rh=0;
        if(root->left) lh=findHeight(root->left, ans);
        if(root->right) rh=findHeight(root->right, ans);

        if(abs(lh-rh)>1) return false;
        return ans;
    }

};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: TreeNode | None) -> bool:

        def dfs(root):
            if root is None:
                return [True,0]

            l_balanced, l_height = dfs(root.left)
            r_balanced, r_height = dfs(root.right)

            is_balanced = l_balanced and r_balanced and abs(l_height-r_height) <= 1
            
            return [is_balanced , 1 + max(l_height,r_height)]

        return dfs(root)[0]